src/embind/embind.js

```diff
@@ -643,7 +637,9 @@ var LibraryEmbind = {
         // Looping here to support possible embedded '0' bytes
         for (var i = 0; i <= length; ++i) {
           var currentBytePtr = value + {{{ POINTER_SIZE }}} + i * charSize;
-          if (i == length || HEAP[currentBytePtr >> shift] == 0) {
+          // Use division instead of shift since the pointer may be greater than
+          // the i32 range.
+          if (i == length || HEAP[currentBytePtr / division] == 0) {
```

Comment:
Ideally we would find a way to only use division on wasm64/CAN_ADDRESS_2GB..

Reply:
I've changed it to use a little helper function that would hopefully be trivial for the js engine to inline. Alternatively, we could add a new preprocessor function like `getHeapOffsetRuntime(offset, size)` where size could be a js variable.